File: backend/app/schemas/collection.py
import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field, model_validator

from backend.app.models.collection import CollectionVisibility
from backend.app.models.gacha import RarityTier
from backend.app.models.relations import MediaEntityType
from backend.app.schemas.relations import EntityRead
from backend.app.utils.media_classification import effective_nsfw_value, effective_sensitive_value


class CollectionMediaRead(BaseModel):
    id: uuid.UUID
    filename: str
    is_nsfw: bool
    is_sensitive: bool
    tags: list[str] = Field(default_factory=list)
    entities: list[EntityRead] = Field(default_factory=list)

    @model_validator(mode="before")
    @classmethod
    def apply_effective_classification(cls, data: Any) -> Any:
        if isinstance(data, dict):
            return data
        if hasattr(data, "is_nsfw") and hasattr(data, "is_sensitive"):
            return {
                "id": data.id,
                "filename": data.filename,
                "is_nsfw": effective_nsfw_value(data),
                "is_sensitive": effective_sensitive_value(data),
                "tags": sorted(mt.tag.name for mt in getattr(data, "media_tags", []) if mt.tag is not None),
                "entities": [
                    entity
                    for entity in getattr(data, "entities", [])
                    if entity.entity_type in {MediaEntityType.character, MediaEntityType.series}
                ],
            }
        return data

    model_config = {"from_attributes": True}


class CollectionItemRead(BaseModel):
    id: uuid.UUID
    user_id: uuid.UUID
    media_id: uuid.UUID
    rarity_tier_at_acquisition: RarityTier
    level: int
    upgrade_xp: int
    copies_pulled: int
    locked: bool
    tradeable: bool
    acquired_at: datetime
    updated_at: datetime
    media: CollectionMediaRead | None = None

    model_config = {"from_attributes": True}


class CollectionItemUpdate(BaseModel):
    locked: bool | None = None
    tradeable: bool | None = None


class CollectionListResponse(BaseModel):
    total: int
    items: list[CollectionItemRead]


class CollectionDiscardResponse(BaseModel):
    item_id: uuid.UUID
    media_id: uuid.UUID
    copies_discarded: int
    pulls_awarded: int
    currency_balance: int
    remaining_copies: int
    item: CollectionItemRead | None = None


class CollectionStatsResponse(BaseModel):
    total_items: int
    total_copies_pulled: int
    duplicate_copies: int
    max_level_items: int
    tier_counts: dict[RarityTier, int]


class CollectionPrivacyRead(BaseModel):
    user_id: uuid.UUID
    visibility: CollectionVisibility
    allow_trade_requests: bool
    show_stats: bool
    show_nsfw: bool

    model_config = {"from_attributes": True}


class CollectionPrivacyUpdate(BaseModel):
    visibility: CollectionVisibility | None = None
    allow_trade_requests: bool | None = None
    show_stats: bool | None = None
    show_nsfw: bool | None = None


class CollectionOwnerRead(BaseModel):
    user_id: uuid.UUID
    username: str
    allow_trade_requests: bool
    show_stats: bool


class CollectionOwnerListResponse(BaseModel):
    total: int
    items: list[CollectionOwnerRead]


class CollectionFilters(BaseModel):
    rarity_tier: RarityTier | None = None
    tags: list[str] | None = None
    character_name: str | None = Field(default=None, max_length=512)
    series_name: str | None = Field(default=None, max_length=512)
    character_names: list[str] | None = None
    series_names: list[str] | None = None
    level: int | None = Field(default=None, ge=1, le=5)
    tradeable: bool | None = None
    duplicates_only: bool = False

File: frontend/src/app/models/collection.ts
import { RarityTier } from './gacha';
import { MediaType } from './media';
import { EntityRead } from './relations';

export enum CollectionVisibility {
  PRIVATE = 'private',
  FOLLOWERS = 'followers',
  PUBLIC = 'public',
}

export interface CollectionMediaRead {
  id: string;
  filename: string;
  media_type?: MediaType;
  is_nsfw: boolean;
  is_sensitive: boolean;
  tags: string[];
  entities: EntityRead[];
}

export interface CollectionItemRead {
  id: string;
  user_id: string;
  media_id: string;
  rarity_tier_at_acquisition: RarityTier;
  level: number;
  upgrade_xp: number;
  copies_pulled: number;
  locked: boolean;
  tradeable: boolean;
  acquired_at: string;
  updated_at: string;
  media: CollectionMediaRead | null;
}

export interface CollectionItemUpdate {
  locked?: boolean | null;
  tradeable?: boolean | null;
}

export interface CollectionListResponse {
  total: number;
  items: CollectionItemRead[];
}

export interface CollectionDiscardResponse {
  item_id: string;
  media_id: string;
  copies_discarded: number;
  pulls_awarded: number;
  currency_balance: number;
  remaining_copies: number;
  item: CollectionItemRead | null;
}

export interface CollectionStatsResponse {
  total_items: number;
  total_copies_pulled: number;
  duplicate_copies: number;
  max_level_items: number;
  tier_counts: Partial<Record<RarityTier, number>>;
}

export interface CollectionPrivacyRead {
  user_id: string;
  visibility: CollectionVisibility;
  allow_trade_requests: boolean;
  show_stats: boolean;
  show_nsfw: boolean;
}

export interface CollectionPrivacyUpdate {
  visibility?: CollectionVisibility | null;
  allow_trade_requests?: boolean | null;
  show_stats?: boolean | null;
  show_nsfw?: boolean | null;
}

export interface CollectionOwnerRead {
  user_id: string;
  username: string;
  allow_trade_requests: boolean;
  show_stats: boolean;
}

export interface CollectionOwnerListResponse {
  total: number;
  items: CollectionOwnerRead[];
}

export interface CollectionOwnerListParams {
  q?: string;
  tradeable_only?: boolean;
}

export interface CollectionListParams {
  rarity_tier?: RarityTier;
  tags?: string[];
  character_name?: string;
  series_name?: string;
  character_names?: string[];
  series_names?: string[];
  level?: number;
  tradeable?: boolean;
  duplicates_only?: boolean;
}
